In [7]:
import random

random_list = [random.random() for _ in range(5_000_000)]

print("Random list length: ", len(random_list))

Random list length:  5000000


In [8]:

total = 0

for i in range(len(random_list)):
    total += random_list[i]
    
total = total / len(random_list)

suma = 0

for i in range(len(random_list)):
    suma += (random_list[i] - total) * (random_list[i] - total)

final = suma / len(random_list)

sd = pow(final, 0.5)

print(sd)

0.2886050253542152


In [9]:
import numpy as np 

arr = np.array(random_list)

print(f'Std: {arr.std()}')


Std: 0.2886050253542154


In [10]:
import pandas as pd
import numpy as np
datos = pd.Series([10, 20, 30, 40, 50])

print("Entorno configurado con exito")
print(datos.describe())

Entorno configurado con exito
count     5.000000
mean     30.000000
std      15.811388
min      10.000000
25%      20.000000
50%      30.000000
75%      40.000000
max      50.000000
dtype: float64


In [3]:
import pandas as pd

url = "https://raw.githubusercontent.com/mricardo89/data-mining/main/Unit-II/datasets/titanic.csv"
df = pd.read_csv(url)

print(df.head(10))

df.describe()


   PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   
5            6         0       3   
6            7         0       1   
7            8         0       3   
8            9         1       3   
9           10         1       2   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   
5                                   Moran, Mr. James    male   NaN      0   
6                            McCarthy, Mr. Timothy J    male  54

,PassengerId,Survived,Pclass,Age,SibSp,Parch,Fare
count,891.000000,891.000000,891.000000,714.000000,891.000000,891.000000,891.000000
mean,446.000000,0.383838,2.308642,29.699118,0.523008,0.381594,32.204208
std,257.353842,0.486592,0.836071,14.526497,1.102743,0.806057,49.693429
min,1.000000,0.000000,1.000000,0.420000,0.000000,0.000000,0.000000
25%,223.500000,0.000000,2.000000,20.125000,0.000000,0.000000,7.910400
50%,446.000000,0.000000,3.000000,28.000000,0.000000,0.000000,14.454200
75%,668.500000,1.000000,3.000000,38.000000,1.000000,0.000000,31.000000
max,891.000000,1.000000,3.000000,80.000000,8.000000,6.000000,512.329200


Pregunta A: Observen la columna Fare (Tarifa del boleto). ¿Cuál es la diferencia matemática entre el percentil 75% (el tercer cuartil) y el valor Máximo (max)? Como ingenieros de software, ¿qué les dice esta enorme dispersión estadística sobre la homogeneidad de los precios que se cobraron en el barco?

R: Se puede ver una diferencia grande entre el valor de los tickets ya que hay ua diferencia de alrededor de 480 unidades de moneda (asumire que son dolares), lo que significa que talvez hubieron diferentes tipos de ventas de boletos, como normales, premium, VIP, VIP+ etc, habiendo un tipo de boleto mucho mas caro que los demas, que solo muy pocos pudieron comprar

Pregunta B: Revisen la fila count (conteo total de registros válidos) de las distintas columnas en el resultado de su .describe(). Comparen el conteo de la columna PassengerId con el conteo de la columna Age. ¿Qué anomalía de base de datos acaban de descubrir?

R: Hay algo raro que se nota ya que en teoria deberian de ser las mismas, ya que cada pasajero tiene su id asignado y por ende tiene vinculado una edad para cada id, y tuvo que haber habido alguna situacion externa para que contara edades de menos a comparacion al id

Pregunta C: El algoritmo que construirán al final del semestre requiere que todos los pasajeros tengan un valor numérico en la columna Age para poder compilar. Sabiendo lo que acaban de descubrir en la Pregunta B, propongan al menos dos estrategias a nivel de base de datos o de programación para resolver este problema antes de alimentar a la IA.

R: Una posibilidad que se me ocurre es que simplemente no usemos valores de la edad para alimentar a la IA, aunque si me parece muy extra deshacernos de una columna de informacion entera, pero talvez quitandola se quita ese ruido del modelo.

Otra posibilidad puede ser rellenar valores con valores basados en otros como medios para rellenar esos valores vacios que no existen en la base de datos, asi se pueden tener esos valores completos sin afectar los resultados finales sin un resultado grave.


In [33]:

import pandas as pd

url = "https://raw.githubusercontent.com/mricardo89/data-mining/main/Unit-II/datasets/titanic.csv"
df = pd.read_csv(url)

totalM = 0
totalF = 0
countM = 0
countF = 0

for i in df.itertuples():
    if i.Sex == "male":
        totalM += i.Fare
        countM += 1
    else:
        totalF += i.Fare
        countF += 1
        
totalM = totalM / countM
totalF = totalF / countF

print(f'La tarifa promedio de los hombres fue {totalM} y el de las mujeres fue {totalF}')  
  
  
df.groupby('Sex')['Fare'].mean()

Q1 = df['Fare'].quantile(0.25)
Q3 = df['Fare'].quantile(0.75)

IQR = Q3 - Q1

min = (Q1) - (1.5*(IQR))
max = (Q3) + (1.5*(IQR))
print(min)
print(max)

maxOutliers = 0
minOutliers = 0


for i in df.itertuples():
    if i.Fare > max:
        maxOutliers += 1
    elif i.Fare < min:
        minOuliers += 1

print(f'Low Outliers: {minOutliers}')
print(f'High Outliers: {maxOutliers}')



La tarifa promedio de los hombres fue 25.523893414211418 y el de las mujeres fue 44.47981783439487
-26.724
65.6344
Low Outliers: 0
High Outliers: 116


In [2]:
import pandas as pd

url = "https://raw.githubusercontent.com/mricardo89/data-mining/main/Unit-II/datasets/titanic.csv"
df = pd.read_csv(url)

print(df['Survived'].value_counts(normalize=True))

porcentajeSurvived = df.groupby('Sex')['Survived'].mean()

print(f'El porcentaje de ambos sexos fue: {porcentajeSurvived}')


Q1 = df['Fare'].quantile(0.25)
Q3 = df['Fare'].quantile(0.75)

IQR = Q3 - Q1

max = Q3 + (1.5 * IQR)

topOutliers = df[df['Fare'] > max]

print(f'El Q1 fue: {Q1}, Q3: {Q3}, IQR: {IQR}, Limite superior aceptable: {max}, el total de outliers: {len(topOutliers)} y por ultimo la distribuicion por clase de los outliers: {topOutliers['Pclass'].value_counts()}')


media = df['Fare'].mean()
mediana = df['Fare'].median()

print(f'La media fue {media} y la mediana fue {mediana}')


nSobrevivieron = df[df['Survived'] == 0]
sobrevivieron = df[df['Survived'] == 1]

n_NSobrevivieron = round(150 * (len(nSobrevivieron) / len(df)))
n_Sobrevivieron = 150 - n_NSobrevivieron

muestra0 = nSobrevivieron.sample(n=n_NSobrevivieron, random_state=42)
muestra1 = sobrevivieron.sample(n=n_Sobrevivieron, random_state=42)

muestra_final = pd.concat([muestra0, muestra1])

print(f"Total en muestra: {len(muestra_final)}")
print(muestra_final['Survived'].value_counts(normalize=True))


print(df.groupby('Survived')['Age'].mean())


varianza_survived = df['Survived'].var()
print(f"Varianza de Survived: {varianza_survived}")

grupos = df.groupby(['Pclass', 'Sex', 'Embarked'])['PassengerId'].count()
print(grupos)


Survived
0    0.616162
1    0.383838
Name: proportion, dtype: float64
El porcentaje de ambos sexos fue: Sex
female    0.742038
male      0.188908
Name: Survived, dtype: float64
El Q1 fue: 7.9104, Q3: 31.0, IQR: 23.0896, Limite superior aceptable: 65.6344, el total de outliers: 116 y por ultimo la distribuicion por clase de los outliers: Pclass
1    104
3      7
2      5
Name: count, dtype: int64
La media fue 32.204207968574636 y la mediana fue 14.4542
Total en muestra: 150
Survived
0    0.613333
1    0.386667
Name: proportion, dtype: float64
Survived
0    30.626179
1    28.343690
Name: Age, dtype: float64
Varianza de Survived: 0.23677221654749742
Pclass  Sex     Embarked
1       female  C            43
                Q             1
                S            48
        male    C            42
                Q             1
                S            79
2       female  C             7
                Q             2
                S            67
        male    C            10



Pregunta A (Sumarización Categórica): Usen la función .value_counts(normalize=True) en la columna Survived (0 = Murió, 1 = Sobrevivió). ¿Cuál es la tasa de supervivencia global del barco expresada en porcentaje?

Survived
0    0.616162
1    0.383838

Esto significa que del total sobrevivieron el 38.38% de las personas a bordo

--------------------------------------------

Pregunta B (Agrupación y Agregación): El famoso código marítimo era "mujeres y niños primero". Comprobemos esto matemáticamente. Ejecuten una agrupación por la columna de género y calculen el promedio de la columna Survived. ¿Qué porcentaje exacto de mujeres sobrevivió en contraste con los hombres?


El porcentaje de ambos sexos fue: Sex
female    0.742038
male      0.188908

Significa que las mujeres si sobrevivieron mas que los hombres

--------------------------------------------

Pregunta C: El director financiero nota que la tarifa máxima (Fare) cobrada fue de más de 500 libras, mientras que el promedio ronda las 32. Sospecha de outliers. Utilicen las herramientas de dispersión en Pandas para confirmarlo:
Calculen el Cuartil 1 (25%) y el Cuartil 3 (75%) de la columna Fare usando el método .quantile([0.25, 0.75]).
Calculen matemáticamente el Rango Intercuartílico (IQR = Q3 - Q1).
Calculen el límite superior aceptable (Q3 + 1.5 * IQR). Escriban una línea de código para filtrar el DataFrame y descubrir exactamente cuántos pasajeros pagaron una tarifa por encima de ese límite matemático. ¿A qué clase (Pclass) pertenecían la mayoría de ellos?

El Q1 fue: 7.9104, Q3: 31.0, IQR: 23.0896, Limite superior aceptable: 65.6344, el total de outliers: 116 y por ultimo la distribuicion por clase de los outliers: Pclass
1    104
3      7
2      5

Gracias a este resultado podemos ver que los datos escenciales (entre el 25% y 75%) estan entre 7.9 y 31, teniendo un IQR de 23.08, y tomando en cuenta esto se calculo el limite superior para considerar a los superiores como outliers de 65.6344, siendo una gran cantidad de 116, tambien viendo que estaban la gran mayoria en el grupo 1 de los que pagaron mas de lo normal, habiendo pocos en el segundo y tercer grupo

--------------------------------------------

Pregunta D: Calculen la media y la mediana de la columna Fare. Notarán una diferencia enorme entre ambos valores. Matemáticamente, ¿qué significa que la media sea tan superior a la mediana? Si en el futuro utilizamos un algoritmo basado en distancias euclidianas (como K-Nearest Neighbors) sin escalar previamente esta variable, ¿cómo afectará esta asimetría al aprendizaje del modelo?

La media fue 32.204207968574636 y la mediana fue 14.4542

Con esto podemos ver que se vio una mediana mucho mas baja que la media considerando que esta por debajo de la mitad del precio, pero claro todos los precios superiores que son muchas veces mas grandes pudieron haber afectado al valor de la media tomando en cuenta esos outliers, clientes o pasageros se verian muy diferente si no se tomaran en cuenta otras variables y solamente lo que pago. Como KNN se basa en ese tipo de distancia, si no se escalan los datos unas columnas pueden dominar demas como Fare, teniendo rangos muy grandes a comparacion de los otros.

--------------------------------------------

Pregunta E: En la semana 2 vimos que el muestreo aleatorio simple es peligroso. El objetivo es entrenar un modelo que prediga la supervivencia (Survived), pero las clases están desbalanceadas. Escriban el código en Pandas para extraer una muestra de exactamente 150 pasajeros garantizando que la proporción de sobrevivientes y no sobrevivientes en la muestra sea idéntica a la de la base de datos completa. ¿Qué sesgo evitan al hacer esto?

Total en muestra: 150
Survived
0    0.613333
1    0.386667

Aqui se evita el sesgo, ya que con el aleatorio simple pudieramos haber terminado con muchos sobrevivientes, o vice versa, entonces no logra identificar correctamente esta variable y puede llegar a alterar de manera mas marcada el resultado que se busca


--------------------------------------------

Pregunta F: Si ejecutan df.groupby('Survived')['Age'].mean(), Pandas calculará el promedio de edad de los que vivieron y los que murieron. Sin embargo, por defecto, Pandas ignora los valores NaN al calcular la media. Si resulta que la gran mayoría de las edades faltantes (NaN) pertenecían a pasajeros de 3ra clase que murieron, ¿qué sesgo estadístico estamos introduciendo involuntariamente en el resultado de esa función y cómo afectaría la inferencia de nuestro modelo?

Se introduciria un sesgo de supervivencia con los datos que faltan, ignorando NAN de dicha clase que fallecio, estariamos calculando solamente del promedio que si se tienen registros completos


--------------------------------------------

Pregunta G: Con el cálculo del IQR en la Pregunta C, determinaron que los boletos de 512 libras son outliers matemáticos. En Machine Learning, los valores atípicos pueden representar errores de captura (ruido que aumenta el error irreducible) o casos especiales válidos (señal). Investigando la naturaleza de un barco de lujo, ¿deberíamos eliminar estas filas con .drop() antes de entrenar nuestro modelo? Justifiquen su respuesta arquitectónica.

Diria que no se deben de eliminar, ya que realmente no es como que representen datos erroneos, son un dato real que debe ser tomado en cuenta, aunque sea un outlier y tenga peso en el resultado final, este dato singular dice mucho sobre precios que pueden llegar a pagar pasajeros de una clase de lujo

--------------------------------------------

Pregunta H: Observen la columna Name. Es texto libre (dato no estructurado), pero contiene títulos ocultos como "Mr.", "Mrs.", "Miss." o "Master.". Si lograran extraer ese título usando expresiones regulares en Pandas, podrían hacer un .groupby('Titulo')['Age'].median(). ¿Por qué imputar las edades faltantes basándose en la mediana del "Título" (ej. "Master" = niño, "Mr" = adulto) sería estadísticamente superior y reduciría el error de nuestro futuro modelo, en comparación con usar la mediana global?

Porque imputar las edades hace que se dividan en grupos estas mismas, ya que se mantiene la correlacion real de las edades de los pasajeros, esto reduciria el error asi mejorando la capacidad predictiva del modelo, teniendo ahora menos rangos de edades por las cuales preocuparse.


--------------------------------------------

Pregunta I: Calculen la varianza de la columna Survived. Dado que es una variable categórica codificada como 0 y 1, el resultado numérico estará cerca de $0.23$. Matemáticamente, ¿qué significaría si la varianza de esta variable fuera exactamente 0.0? ¿Qué pasaría si intentan entrenar un algoritmo de clasificación con un dataset donde la variable de respuesta tiene varianza 0.0?

La varianza obtenida fue: 0.23677221654749742

Si la varianza hubiera sido 0, hubiera significado que no habia ninguna variacion como tal en los datos, osea que todos los pasajeros hubieran tenido el mismo resultado, 1 o 0 (vivo o muerto).
Si intentamos entrenar algo con 0.0 el algoritmo no serviria ya que no tendria cosas que comparar para aprender a clasificarlo, dando el mismo resultado siempre

--------------------------------------------

Pregunta J: Ejecuten una agrupación por tres niveles al mismo tiempo y cuenten cuántos pasajeros hay en cada subgrupo: df.groupby(['Pclass', 'Sex', 'Embarked'])['PassengerId'].count(). Notarán que algunos subgrupos tienen 1 o 2 pasajeros. Si un algoritmo intenta extraer reglas de probabilidad de grupos tan pequeños, se enfrentará a la "Maldición de la Dimensionalidad" (Curse of Dimensionality). ¿Qué fenómeno perjudicial (sobreajuste o subajuste) ocurrirá inevitablemente si dejamos que el modelo aprenda reglas basadas en esos grupos de 1 solo pasajero?


Este fue el resultado:
Pclass  Sex     Embarked
1       female  C            43
                Q             1
                S            48
        male    C            42
                Q             1
                S            79
2       female  C             7
                Q             2
                S            67
        male    C            10
                Q             1
                S            97
3       female  C            23
                Q            33
                S            88
        male    C            43
                Q            39
                S           265

Lo que pasaria es el overfitting, ya que al tener grupos de 1 o 2 personas, este modelo no aprenderia de algun tipo de patron, solamente entenderia lo que se le esta dando, asi memorizandoselo ya que no hay variacion en los datos. Entonces este funcionaria completamente correcto con los datos de practica, pero en un caso real no seria para nada bueno ya que se acostumbro a los datos de entrenamiento demasiado.
